File: packages/polars/lib/src/str.dart
import '../polars.dart';

/// Sugar for `Expr.str*` methods.
///
/// See [Strings](https://docs.pola.rs/user-guide/expressions/strings) for more details.
///
/// Many functions in this namespace accept a `pat` parameter, which is parsed
/// as a regular expression except when the parameter `literal` is set to `true`.
class StrNamespace {
  final Expr expr;
  const StrNamespace(this.expr);

  Expr containsLiteral(Object? pat) => expr.strContainsLiteral(pat: pat.expr);
  Expr contains(Object? pat, {bool strict = true}) =>
      expr.strContains(pat: pat.expr, strict: strict);
  Expr endsWith(Object? pat) => expr.strEndsWith(pat: pat.expr);
  Expr extract(String pat, {required int groupIndex}) =>
      expr.strExtract(pat: pat, groupIndex: groupIndex);
  Expr extractAll(Object? pat) => expr.strExtractAll(pat: pat.expr);
  Expr countMatches(Object? pat, {bool literal = false}) =>
      expr.strCountMatches(pat: pat.expr, literal: literal);
  Expr concat(String delimiter, {bool ignoreNulls = true}) =>
      expr.strConcat(delimiter: delimiter, ignoreNulls: ignoreNulls);
  Expr splitN({required Expr by, required int n}) =>
      expr.strSplitn(by: by, n: n);
  Expr replace(Object? pat, Object? val, {bool literal = false}) =>
      expr.strReplace(pat: pat.expr, val: val.expr, literal: literal);
  Expr replaceN(
    Object? pat,
    Object? val, {
    required int n,
    bool literal = false,
  }) =>
      expr.strReplaceN(pat: pat.expr, val: val.expr, n: n, literal: literal);
  Expr replaceAll(Object? pat, Object? val, {bool literal = false}) =>
      expr.strReplaceAll(pat: pat.expr, val: val.expr, literal: literal);
  Expr stripChars(Object? matches) => expr.strStripChars(matches: matches.expr);
  Expr stripCharsStart(Object? matches) =>
      expr.strStripCharsStart(matches: matches.expr);
  Expr stripCharsEnd(Object? matches) =>
      expr.strStripCharsEnd(matches: matches.expr);
  Expr stripPrefix(Object? prefix) => expr.strStripPrefix(prefix: prefix.expr);
  Expr stripSuffix(Object? suffix) => expr.strStripSuffix(suffix: suffix.expr);
  Expr get toLowercase => expr.strToLowercase();
  Expr get toUppercase => expr.strToUppercase();
  Expr toInteger({required int base, bool strict = true}) =>
      expr.strToInteger(base: base, strict: strict);
  Expr get lenBytes => expr.strLenBytes();
  Expr get lenChars => expr.strLenChars();
  Expr slice(int start, int? length) =>
      expr.strSlice(start: start, length: length);
  Expr get explode => expr.strExplode();

  Expr strptime(
    DataType dtype, {
    String? format,
    bool strict = true,
    bool exact = true,
    bool cache = true,
    Ambiguous ambiguous = Ambiguous.raise,
  }) =>
      expr.strptime(
        dtype: dtype,
        format: format,
        strict: strict,
        exact: exact,
        cache: cache,
        ambiguous: ambiguous,
      );

  Expr toDate({
    String? format,
    bool strict = true,
    bool exact = true,
    bool cache = true,
  }) =>
      expr.strToDate(
        format: format,
        strict: strict,
        exact: exact,
        cache: cache,
      );

  Expr toDatetime({
    String? format,
    TimeUnit? timeUnit,
    String? timeZone,
  }) =>
      expr.strToDatetime(
        format: format,
        timeUnit: timeUnit,
        timeZone: timeZone,
      );

  Expr toTime({
    String? format,
    bool strict = true,
    bool exact = true,
    bool cache = true,
  }) =>
      expr.strToTime(
        format: format,
        strict: strict,
        exact: exact,
        cache: cache,
      );

  Expr split({
    required Object? by,
    bool inclusive = false,
  }) =>
      expr.strSplit(by: by.expr, inclusive: inclusive);

  Expr splitExact({
    required Object? by,
    required int n,
    bool inclusive = false,
  }) =>
      expr.strSplitExact(by: by.expr, n: n, inclusive: inclusive);
}

File: polars-wrapper/src/wrapper/str.rs
use flutter_rust_bridge::frb;
use polars::{datatypes::TimeUnit, lazy::dsl::StrptimeOptions};

use super::{
    df::DataType,
    expr::{Ambiguous, Expr},
};

macro_rules! delegate_str {
    ($( $(#[$attribute:meta])* $fn:ident(self $(,)? $($param:ident : $(#[$conv:ident])? $ty:ty $(= $default:expr)? ),*) -> $output:ty; )*) => {paste::paste!{$(
        $(#[$attribute])*
        #[doc = concat!(" TODO: Docs for ", stringify!($fn))]
        pub fn [<str_ $fn>](&self, $($(#[frb(default = $default)])? $param : $ty),*) -> $output {
            <$output>::from(self.into_internal().str().$fn($($param $(.$conv())?),*))
        }
    )*}};
}
impl Expr {
    delegate_str! {
        #[frb(sync)] contains_literal(self, pat: #[into] Expr) -> Expr;
        #[frb(sync)] contains(self, pat: #[into] Expr, strict: bool = true) -> Expr;
        #[frb(sync)] ends_with(self, pat: #[into] Expr) -> Expr;
        #[frb(sync)] extract(self, pat: #[as_str] String, group_index: usize) -> Expr;
        #[frb(sync)] extract_all(self, pat: #[into] Expr) -> Expr;
        #[frb(sync)] count_matches(self, pat: #[into] Expr, literal: bool = false) -> Expr;
        #[frb(sync)] concat(self, delimiter: #[as_str] String, ignore_nulls: bool = true) -> Expr;
        #[frb(sync)] splitn(self, by: #[into] Expr, n: usize) -> Expr;
        #[frb(sync)] replace(self, pat: #[into] Expr, val: #[into] Expr, literal: bool = false) -> Expr;
        #[frb(sync)] replace_n(self, pat: #[into] Expr, val: #[into] Expr, literal: bool = false, n: i64) -> Expr;
        #[frb(sync)] replace_all(self, pat: #[into] Expr, val: #[into] Expr, literal: bool = false) -> Expr;
        #[frb(sync)] starts_with(self, pat: #[into] Expr) -> Expr;
        #[frb(sync)] strip_chars(self, matches: #[into] Expr) -> Expr;
        #[frb(sync)] strip_chars_start(self, matches: #[into] Expr) -> Expr;
        #[frb(sync)] strip_chars_end(self, matches: #[into] Expr) -> Expr;
        #[frb(sync)] strip_prefix(self, prefix: #[into] Expr) -> Expr;
        #[frb(sync)] strip_suffix(self, suffix: #[into] Expr) -> Expr;
        #[frb(sync, getter)] to_lowercase(self) -> Expr;
        #[frb(sync, getter)] to_uppercase(self) -> Expr;
        #[frb(sync)] to_integer(self, base: u32, strict: bool = true) -> Expr;
        #[frb(sync, getter)] len_bytes(self) -> Expr;
        #[frb(sync, getter)] len_chars(self) -> Expr;
        #[frb(sync)] slice(self, start: i64, length: Option<u64>) -> Expr;
        #[frb(sync, getter)] explode(self) -> Expr;
    }
    /// - `dtype` A temporal data type, i.e. Date, DateTime, or Time.
    #[frb(sync)]
    pub fn strptime(
        &self,
        dtype: DataType,
        format: Option<String>,
        #[frb(default = true)] strict: bool,
        #[frb(default = true)] exact: bool,
        #[frb(default = true)] cache: bool,
        #[frb(default = "Ambiguous.raise")] ambiguous: Ambiguous,
    ) -> Expr {
        Expr::from(self.into_internal().str().strptime(
            dtype.into(),
            StrptimeOptions {
                format,
                strict,
                exact,
                cache,
            },
            ambiguous.into_expr(),
        ))
    }
    #[frb(sync)]
    pub fn str_to_date(
        &self,
        format: Option<String>,
        #[frb(default = true)] strict: bool,
        #[frb(default = true)] exact: bool,
        #[frb(default = true)] cache: bool,
    ) -> Expr {
        Expr::from(self.into_internal().str().to_date(StrptimeOptions {
            format,
            strict,
            exact,
            cache,
        }))
    }
    #[frb(sync)]
    pub fn str_to_datetime(
        &self,
        time_unit: Option<TimeUnit>,
        time_zone: Option<String>,
        format: Option<String>,
        #[frb(default = true)] strict: bool,
        #[frb(default = true)] exact: bool,
        #[frb(default = true)] cache: bool,
        #[frb(default = "Ambiguous.raise")] ambiguous: Ambiguous,
    ) -> Expr {
        Expr::from(self.into_internal().str().to_datetime(
            time_unit,
            time_zone,
            StrptimeOptions {
                format,
                strict,
                exact,
                cache,
            },
            ambiguous.into_expr(),
        ))
    }
    #[frb(sync)]
    pub fn str_to_time(
        &self,
        format: Option<String>,
        #[frb(default = true)] strict: bool,
        #[frb(default = true)] exact: bool,
        #[frb(default = true)] cache: bool,
    ) -> Expr {
        Expr::from(self.into_internal().str().to_time(StrptimeOptions {
            format,
            strict,
            exact,
            cache,
        }))
    }
    #[frb(sync)]
    pub fn str_split(&self, by: Expr, #[frb(default = false)] inclusive: bool) -> Expr {
        let str = self.into_internal().str();
        if inclusive {
            Expr::from(str.split_inclusive(by.into()))
        } else {
            Expr::from(str.split(by.into()))
        }
    }
    #[frb(sync)]
    pub fn str_split_exact(
        &self,
        by: Expr,
        n: usize,
        #[frb(default = false)] inclusive: bool,
    ) -> Expr {
        let str = self.into_internal().str();
        if inclusive {
            Expr::from(str.split_exact_inclusive(by.into(), n))
        } else {
            Expr::from(str.split_exact(by.into(), n))
        }
    }
}
